File: lang/typescript/lib/BaseIfc.ts
export abstract class BaseIfc
{
    private _id : string
    private _stepId : string

    constructor() {
        this._id = "XXXXX"
    }

    get id() : string {
        return this._id
    }

    get stepId() :string {
        return this._stepId
    }

    set stepId(value: string) {
        this._stepId = value
    }

    toJSON() : string {
        return JSON.stringify(this)
    }

    toSTEP() : string {
        let ifcClass : string = this.constructor.name.toUpperCase();
        return `${this.stepId} = ${ifcClass}(${this.getStepParameters()})`
    }

    toStepValue(value: any, isSelectOption : boolean = false) : string {
        return `${this.stepId}`
    }

    getStepParameters() : string {
        return ""
    }
}

File: lang/csharp/src/IFC/BaseIfc.cs
using System;
using Newtonsoft.Json;
using STEP;

namespace IFC
{
    public abstract class BaseIfc : IConvertibleToSTEP
    {
        [JsonProperty("id")]
        public Guid Id { get; }

        public int StepId { get; set; }

        public BaseIfc()
        {
            Id = Guid.NewGuid();
        }

        public virtual string ToJSON()
        {
            var settings = new JsonSerializerSettings()
            {
                Formatting = Formatting.Indented,
                TypeNameHandling = TypeNameHandling.Objects
            };
            return JsonConvert.SerializeObject(this, settings);
        }

        public virtual string ToSTEP()
        {
            string IfcClass = GetType().Name.ToUpper();
            return $"#{StepId} = {IfcClass}({this.GetStepParameters()});";
        }

        public virtual string ToStepValue(bool isSelectOption = false)
        {
            return $"#{StepId}";
        }

        public virtual string GetStepParameters()
        {
            return string.Empty;
        }
    }
}